### PR TITLE
[SYCLomatic] Integrate pattern-rewriter into user-defined rule

### PR DESCRIPTION
**Pull request by AndyCHHuang**

### CODE REVIEW

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/lib/DPCT/Rules.cpp [hunk 1]
```diff
@@ -654,3 +670,496 @@ void clang::dpct::UserDefinedEnumRule::runRule(
     EA.applyAllSubExprRepl();
   }
 }
+
+namespace pattern_rewriter{
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Rules.cpp]: can we reorg the file to put pattern rewritten code into separate new file.

File: clang/test/dpct/user_defined_rule.cu [hunk 2]
```diff
@@ -158,4 +158,14 @@ __device__ void foo5(){ OldType<int> *cu_st;}
 __global__ void foo6(){
 //CHECK: std::ceil(2.5);
   std::ceil(2.5);
+}
+
+int my_max(int a, int b) {
+  return 0;
```
Comment (zhimingwang36) [inline clang/test/dpct/user_defined_rule.cu:164]: Let's add readme in extensions/opt_rules to add guide user to add post processing rules and maybe other user defined rules

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**